synapse/storage/databases/main/relations.py

```diff
@@ -142,31 +144,32 @@ def _get_recent_references_for_event_txn(
         ) -> Tuple[List[_RelatedEvent], Optional[StreamToken]]:
             txn.execute(sql, where_args + [limit + 1])
 
-            last_topo_id = None
-            last_stream_id = None
             events = []
             for event_id, relation_type, sender, topo_ordering, stream_ordering in txn:
                 # Do not include edits for redacted events as they leak event
                 # content.
                 if not is_redacted or relation_type != RelationTypes.REPLACE:
-                    events.append(_RelatedEvent(event_id, sender))
-                last_topo_id = topo_ordering
-                last_stream_id = stream_ordering
+                    events.append(
+                        _RelatedEvent(event_id, sender, topo_ordering, stream_ordering)
+                    )
 
-            # If there are more events, generate the next pagination key.
+            # If there are more events, generate the next pagination key from the
+            # last event returned.
             next_token = None
             if len(events) > limit:
-                # If there are events we must have pulled at least one row.
-                assert last_topo_id is not None
-                assert last_stream_id is not None
-
-                # Due to how the pagination clause is generated, the stream ID is
-                # handled as an exclusive range when paginating forward. Correct
-                # for that here.
-                if direction == "f":
-                    last_stream_id -= 1
+                events = events[:limit]
+
+                topo = events[-1].topological_ordering
+                toke = events[-1].stream_ordering
```

Comment:
Quoting https://github.com/matrix-org/synapse/pull/13840#discussion_r976846353:

> Instead of looking at the _last_ row (which is what tells us there's more data), we want to look at the second to last row.

Took me a moment to see that this happens here, on Line +160.

Reply:
I can add a comment!